operator/builtin/input/file/design.md

```diff
@@ -0,0 +1,146 @@
+
+
+### Persistence
+
+Readers are always instantiated with an open file handle. Eventually, the file handle is closed, but the Reader is not discarded. Rather, it is maintained for a fixed number of "poll cycles" (see Polling section below) as a reference to the file's metadata, which may be useful for detecting files that have been moved or copied, and for recalling metadata such as the file's previous path.
+
+Readers are maintained for a fixed period of time, and then discarded.
+
+When the `file_input` operator makes use of a persistence mechanism to save and recall its state, it is simply Setting and Getting a slice of Readers. These Readers contain all the information necessary to pick up exactly where the operator left off.
+
+
+# Polling
+
+The file system is polled on a regular interval, defined by the `poll_interval` setting. 
+
+Each poll cycle runs through a series of steps which are presented below.
+
+At a very high level, each poll cycle operates as three phases:
+1. Finish work that was started in the previous poll cycle.
+2. Begin work that will carry over to the next cycle.
+3. Allow some time to pass.
+
+Because state is carried over from one poll cycle to the next, the following detailed description is presented starting mid-cycle, at a point where there is no preexisting state.
+
+
+### Detailed Poll Cycle
+
+_Note: The following steps are presented starting mid poll cycle._
+
+1. Matching
+    1. Find files that match the `include` setting. The files are known only by their paths.
+    2. Discard any of these files that match the `exclude` setting.
+    3. As a special case, on the first poll cycle, a warning is printed if no files are matched. Execution continues regardless.
+2. Queueing
+    1. If the number of matched files is less than or equal to the maximum degree of concurrency, as defined by the `max_concurrent_files` setting, then no queueing occurs.
+    2. Else, queueing occurs, which means the following:
+        - Matched files are split into two sets, such that the first is small enough to respect `max_concurrent_files`, and the second contains the remaining files (called the queue).
+        - The current poll interval will begin processing the first set of files, just as if they were the only ones found during the matching phase.
+        - Subsequent poll cycles will pull matches off of the queue, until the queue is empty.
+        - The `max_concurrent_files` setting is respected at all times.
+        - TODO: Use a worker pool to process the queue as quickly as possible, rather than waiting for poll cycles to trigger batching.
+3. Opening
+    1. Each of the matched files is opened. Note:
+        - A small amount of time has passed since the file was matched.
+        - It is possible that is has been moved or deleted by this point.
+        - Only a minimum set of operations should occur between file matching and opening.
+        - If an error occurs while opening, it is logged.
```

Comment:
Does this mean every time a log rotation via renaming happens we are likely to log an error message?

Reply:
Some other rare errors are likely, possible which we would not mind logging.

Those aside, I believe we would only expect to see this if the following is true:
1. Running on Windows
2. Logs going to a file rather than Windows Event Log
3. Logs are being rotated using move/create strategy

I'm sure all of this happens, but in theory the log emitter should make different choices. Of course this isn't always in the control of the user, so perhaps this would be an annoying log in that case.

Do you agree with the above assessment? If so, do you think logging an error is reasonable?

Reply:
> Do you agree with the above assessment? If so, do you think logging an error is reasonable?

I think it is reasonable. We can rethink this in the future if we find that the users get scared from the error messages.